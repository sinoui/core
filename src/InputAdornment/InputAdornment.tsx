import React from 'react';
import styled, { css } from 'styled-components';
import Body1 from '@sinoui/core/Body1';
import bemClassNames from '../utils/bemClassNames';

export interface Props {
  /**
   * 装饰器相对于输入框的位置。
   */
  position: 'start' | 'end';
  /**
   * 禁止点击事件。这个属性的作用是：点击装饰器会让输入框获取到焦点。
   */
  disablePointerEvents?: boolean;
  /**
   * 组件的内容。通常是一个 IconButton 或者字符串。
   */
  children: React.ReactNode;
  /**
   * 根元素使用的组件。默认为 `div`。
   */
  as?: React.ElementType;
  /**
   * 自定义样式类名称
   */
  className?: string;
}

const endCss = css`
  margin-left: 8px;
`;

const startCss = css`
  margin-right: 8px;
`;

interface InputAdornmentLayoutProps {
  /**
   * 装饰器相对于输入框的位置。
   */
  $position: 'start' | 'end';
  /**
   * 禁止点击事件。这个属性的作用是：点击装饰器会让输入框获取到焦点。
   */
  $disablePointerEvents?: boolean;
  $isText?: boolean;
}

const InputAdornmentLayout = styled.div<InputAdornmentLayoutProps>`
  flex-shrink: 0;
  display: flex;
  height: 0.01em;
  align-items: center;
  max-height: 2em;
  white-space: nowrap;
  color: ${({ theme }) => theme.palette.text.secondary};
  [disabled] > & {
    color: ${({ theme }) => theme.palette.text.disabled};
  }

  ${({ $position, $isText }) =>
    !$isText && ($position === 'start' ? startCss : endCss)};
  ${({ $disablePointerEvents }) =>
    $disablePointerEvents &&
    `
      pointer-events: none;
      user-select: none;
  `};
`;

/**
 * 输入框装饰器
 */
const InputAdornment = (props: Props) => {
  const {
    children,
    disablePointerEvents,
    position,
    className,
    ...rest
  } = props;
  const isTextChildren = typeof children === 'string';
  const $disablePointerEvents = disablePointerEvents ?? isTextChildren;

  return (
    <InputAdornmentLayout
      {...rest}
      $disablePointerEvents={$disablePointerEvents}
      $position={position}
      $isText={isTextChildren}
      className={bemClassNames(
        'sinoui-input-adornment',
        {
          start: position === 'start',
          end: position === 'end',
          text: isTextChildren,
        },
        className,
      )}
    >
      {isTextChildren ? <Body1 color="inherit">{children}</Body1> : children}
    </InputAdornmentLayout>
  );
};

export default InputAdornment;
