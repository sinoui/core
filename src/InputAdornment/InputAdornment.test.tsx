import 'jest-styled-components';
import React from 'react';
import renderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';
import { defaultTheme } from '@sinoui/theme';
import InputAdornment from './InputAdornment';

it('在开始位置使用', () => {
  const tree = renderer
    .create(
      <ThemeProvider theme={defaultTheme}>
        <InputAdornment position="start">
          <p>装饰器</p>
        </InputAdornment>
      </ThemeProvider>,
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('在结束位置使用', () => {
  const tree = renderer
    .create(
      <ThemeProvider theme={defaultTheme}>
        <InputAdornment position="end">
          <p>装饰器</p>
        </InputAdornment>
      </ThemeProvider>,
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('显示文本装饰器', () => {
  const tree = renderer
    .create(
      <ThemeProvider theme={defaultTheme}>
        <InputAdornment position="end">装饰器</InputAdornment>
      </ThemeProvider>,
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('禁止点击事件', () => {
  const tree = renderer
    .create(
      <ThemeProvider theme={defaultTheme}>
        <InputAdornment position="start" disablePointerEvents>
          <p>装饰器</p>
        </InputAdornment>
      </ThemeProvider>,
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it('替换根元素', () => {
  const tree = renderer
    .create(
      <ThemeProvider theme={defaultTheme}>
        <InputAdornment position="start" as="p">
          <p>装饰器</p>
        </InputAdornment>
      </ThemeProvider>,
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
