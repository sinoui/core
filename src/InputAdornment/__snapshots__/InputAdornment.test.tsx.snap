// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`在开始位置使用 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 0.01em;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  max-height: 2em;
  white-space: nowrap;
  margin-right: 8px;
}

<div
  className="c0 sinoui-input-adornment"
>
  <p>
    装饰器
  </p>
</div>
`;

exports[`在结束位置使用 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 0.01em;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  max-height: 2em;
  white-space: nowrap;
  margin-left: 8px;
}

<div
  className="c0 sinoui-input-adornment"
>
  <p>
    装饰器
  </p>
</div>
`;

exports[`显示文本装饰器 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 0.01em;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  max-height: 2em;
  white-space: nowrap;
  margin-left: 8px;
  pointer-events: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c1 {
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.5;
  -webkit-letter-spacing: 0.00938em;
  -moz-letter-spacing: 0.00938em;
  -ms-letter-spacing: 0.00938em;
  letter-spacing: 0.00938em;
  color: rgba(0,0,0,0.6);
  margin: 0;
}

<div
  className="c0 sinoui-input-adornment"
>
  <p
    className="c1"
  >
    装饰器
  </p>
</div>
`;

exports[`替换根元素 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 0.01em;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  max-height: 2em;
  white-space: nowrap;
  margin-right: 8px;
}

<p
  className="c0 sinoui-input-adornment"
>
  <p>
    装饰器
  </p>
</p>
`;

exports[`禁止点击事件 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 0.01em;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  max-height: 2em;
  white-space: nowrap;
  margin-right: 8px;
  pointer-events: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

<div
  className="c0 sinoui-input-adornment"
>
  <p>
    装饰器
  </p>
</div>
`;
